crates/goose/src/session/storage.rs

```diff
@@ -20,14 +20,150 @@ use std::fs;
 use std::io::{self, BufRead, Write};
 use std::ops::DerefMut;
 use std::path::{Path, PathBuf};
-use std::sync::Arc;
+use std::sync::{Arc, OnceLock};
+use std::time::{Duration, Instant};
+use tokio::sync::mpsc;
+use tokio::time::sleep;
 use utoipa::ToSchema;
 
 // Security limits
 const MAX_FILE_SIZE: u64 = 10 * 1024 * 1024; // 10MB
 const MAX_MESSAGE_COUNT: usize = 5000;
 const MAX_LINE_LENGTH: usize = 1024 * 1024; // 1MB per line
 
+// Async persistence constants
+const BATCH_SAVE_INTERVAL: Duration = Duration::from_millis(500);
+const MAX_BATCH_SIZE: usize = 10;
+
+/// Messages for the background persistence worker
+enum PersistenceTask {
+    Save {
+        session_file: PathBuf,
+        messages: Conversation,
+        metadata: Box<SessionMetadata>,
+    },
+    GenerateName {
+        session_file: PathBuf,
+        messages: Conversation,
+        metadata: Box<SessionMetadata>,
+        provider: Arc<dyn Provider>,
+    },
+}
+
+/// Global background persistence worker
+static SESSION_SAVER: OnceLock<mpsc::UnboundedSender<PersistenceTask>> = OnceLock::new();
+static _SESSION_SAVER_TASK: OnceLock<tokio::task::JoinHandle<()>> = OnceLock::new();
+
+/// Initialize the background session saver
+fn init_session_saver() -> mpsc::UnboundedSender<PersistenceTask> {
+    let (tx, mut rx) = mpsc::unbounded_channel::<PersistenceTask>();
+
+    let task_handle = tokio::spawn(async move {
+        let mut batch: Vec<PersistenceTask> = Vec::new();
+        let mut last_save = Instant::now();
+
+        loop {
+            // Try to receive tasks with a timeout
+            match tokio::time::timeout(BATCH_SAVE_INTERVAL, rx.recv()).await {
+                Ok(Some(task)) => {
+                    batch.push(task);
+
+                    // Save batch if it's full or enough time has passed
+                    if batch.len() >= MAX_BATCH_SIZE || last_save.elapsed() >= BATCH_SAVE_INTERVAL {
+                        save_batch(&mut batch).await;
+                        last_save = Instant::now();
+                    }
+                }
+                Ok(None) => break, // Channel closed
+                Err(_) => {
+                    // Timeout - save any pending tasks
+                    if !batch.is_empty() {
+                        save_batch(&mut batch).await;
+                        last_save = Instant::now();
+                    }
+                }
+            }
+        }
+
+        // Save any remaining tasks before exiting
+        if !batch.is_empty() {
+            save_batch(&mut batch).await;
+        }
+    });
+
+    // Store the task handle to prevent it from being dropped
+    let _ = _SESSION_SAVER_TASK.set(task_handle);
+    tx
+}
+
+/// Save a batch of persistence tasks
+async fn save_batch(batch: &mut Vec<PersistenceTask>) {
```

Comment:
Most of the time we'll be queueing saves to the same file, but it wouldn't make sense to batch them, we should just drop all but the latest one, right?